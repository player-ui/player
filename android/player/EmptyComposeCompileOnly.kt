// This file is intentionally left empty to allow the compose_compileonly target to have deps and neverlink. 