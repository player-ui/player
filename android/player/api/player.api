public final class com/intuit/player/android/AndroidPlayer : com/intuit/player/jvm/core/player/Player {
	public fun <init> (Ljava/util/List;)V
	public fun <init> ([Lcom/intuit/player/jvm/core/plugins/Plugin;)V
	public final fun commitPendingTransaction ()V
	public final fun getCachedStyledContext (Landroid/content/Context;Ljava/util/List;)Landroid/content/Context;
	public fun getHooks ()Lcom/intuit/player/android/AndroidPlayer$Hooks;
	public synthetic fun getHooks ()Lcom/intuit/player/jvm/core/player/Player$Hooks;
	public fun getLogger ()Lcom/intuit/player/jvm/core/logger/TapableLogger;
	public fun getPlugins ()Ljava/util/List;
	public fun getState ()Lcom/intuit/player/jvm/core/player/state/PlayerFlowState;
	public final fun onUpdate (Lkotlin/jvm/functions/Function2;)V
	public final fun recycle ()V
	public final fun registerAsset (Lkotlin/reflect/KClass;Ljava/lang/String;Lkotlin/jvm/functions/Function1;)V
	public final fun registerAsset (Lkotlin/reflect/KClass;Ljava/util/Map;Lkotlin/jvm/functions/Function1;)V
	public final fun registerPendingTransaction (ZLcom/intuit/player/android/AndroidPlayer$PendingTransaction;)V
	public static synthetic fun registerPendingTransaction$default (Lcom/intuit/player/android/AndroidPlayer;ZLcom/intuit/player/android/AndroidPlayer$PendingTransaction;ILjava/lang/Object;)V
	public fun release ()V
	public fun start (Ljava/lang/String;)Lcom/intuit/player/jvm/core/bridge/Completable;
}

public final class com/intuit/player/android/AndroidPlayer$Hooks : com/intuit/player/jvm/core/player/Player$Hooks {
	public final fun getContext ()Lcom/intuit/player/android/AndroidPlayer$Hooks$ContextHook;
	public fun getDataController ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getExpressionEvaluator ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getFlowController ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getState ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getView ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
	public fun getViewController ()Lcom/intuit/player/jvm/core/bridge/hooks/NodeSyncHook1;
}

public final class com/intuit/player/android/AndroidPlayer$Hooks$ContextHook : com/intuit/hooks/SyncWaterfallHook {
	public fun <init> ()V
	public final fun call (Landroid/content/Context;)Landroid/content/Context;
}

public abstract interface class com/intuit/player/android/AndroidPlayer$PendingTransaction {
	public abstract fun commit ()V
}

public abstract interface class com/intuit/player/android/AndroidPlayerPlugin : com/intuit/player/jvm/core/plugins/Plugin {
	public abstract fun apply (Lcom/intuit/player/android/AndroidPlayer;)V
}

public final class com/intuit/player/android/AssetContext {
	public fun <init> (Landroid/content/Context;Lcom/intuit/player/jvm/core/asset/Asset;Lcom/intuit/player/android/AndroidPlayer;Lkotlin/jvm/functions/Function1;)V
	public fun <init> (Landroid/content/Context;Lcom/intuit/player/jvm/core/asset/Asset;Lcom/intuit/player/android/AndroidPlayer;Lkotlin/jvm/functions/Function1;Ljava/lang/String;)V
	public final fun component1 ()Landroid/content/Context;
	public final fun component2 ()Lcom/intuit/player/jvm/core/asset/Asset;
	public final fun component3 ()Lcom/intuit/player/android/AndroidPlayer;
	public final fun component5 ()Ljava/lang/String;
	public final fun copy (Landroid/content/Context;Lcom/intuit/player/jvm/core/asset/Asset;Lcom/intuit/player/android/AndroidPlayer;Lkotlin/jvm/functions/Function1;Ljava/lang/String;)Lcom/intuit/player/android/AssetContext;
	public static synthetic fun copy$default (Lcom/intuit/player/android/AssetContext;Landroid/content/Context;Lcom/intuit/player/jvm/core/asset/Asset;Lcom/intuit/player/android/AndroidPlayer;Lkotlin/jvm/functions/Function1;Ljava/lang/String;ILjava/lang/Object;)Lcom/intuit/player/android/AssetContext;
	public fun equals (Ljava/lang/Object;)Z
	public final fun getAsset ()Lcom/intuit/player/jvm/core/asset/Asset;
	public final fun getContext ()Landroid/content/Context;
	public final fun getId ()Ljava/lang/String;
	public final fun getPlayer ()Lcom/intuit/player/android/AndroidPlayer;
	public final fun getType ()Ljava/lang/String;
	public fun hashCode ()I
	public final fun postFixId (Ljava/lang/String;)Lcom/intuit/player/android/AssetContext;
	public fun toString ()Ljava/lang/String;
}

public final class com/intuit/player/android/AssetContextKt {
	public static final fun build (Lcom/intuit/player/android/AssetContext;)Lcom/intuit/player/android/asset/RenderableAsset;
	public static final fun withContext (Lcom/intuit/player/android/AssetContext;Landroid/content/Context;)Lcom/intuit/player/android/AssetContext;
	public static final fun withStyles (Lcom/intuit/player/android/AssetContext;Ljava/util/List;)Lcom/intuit/player/android/AssetContext;
	public static final fun withStyles (Lcom/intuit/player/android/AssetContext;[Ljava/lang/Integer;)Lcom/intuit/player/android/AssetContext;
	public static final fun withTag (Lcom/intuit/player/android/AssetContext;Ljava/lang/String;)Lcom/intuit/player/android/AssetContext;
}

public final class com/intuit/player/android/BuildConfig {
	public static final field BUILD_TYPE Ljava/lang/String;
	public static final field DEBUG Z
	public static final field LIBRARY_PACKAGE_NAME Ljava/lang/String;
	public fun <init> ()V
}

public abstract class com/intuit/player/android/asset/DecodableAsset : com/intuit/player/android/asset/RenderableAsset {
	public fun <init> (Lcom/intuit/player/android/AssetContext;Lkotlinx/serialization/KSerializer;)V
	public final fun getData ()Ljava/lang/Object;
}

public abstract class com/intuit/player/android/asset/RenderableAsset : com/intuit/player/jvm/core/bridge/NodeWrapper {
	public fun <init> (Lcom/intuit/player/android/AssetContext;)V
	public final fun beacon (Ljava/lang/String;Ljava/lang/String;Lcom/intuit/player/jvm/core/asset/Asset;Ljava/lang/Object;)V
	public static synthetic fun beacon$default (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/lang/String;Ljava/lang/String;Lcom/intuit/player/jvm/core/asset/Asset;Ljava/lang/Object;ILjava/lang/Object;)V
	public final fun expand (Lcom/intuit/player/jvm/core/asset/AssetWrapper;Landroid/content/Context;)Lcom/intuit/player/android/asset/RenderableAsset;
	public final fun expand (Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;Landroid/content/Context;)Lcom/intuit/player/android/asset/RenderableAsset;
	public final fun expand (Ljava/lang/String;Landroid/content/Context;)Lcom/intuit/player/android/asset/RenderableAsset;
	public static synthetic fun expand$default (Lcom/intuit/player/android/asset/RenderableAsset;Lcom/intuit/player/jvm/core/asset/AssetWrapper;Landroid/content/Context;ILjava/lang/Object;)Lcom/intuit/player/android/asset/RenderableAsset;
	public static synthetic fun expand$default (Lcom/intuit/player/android/asset/RenderableAsset;Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;Landroid/content/Context;ILjava/lang/Object;)Lcom/intuit/player/android/asset/RenderableAsset;
	public static synthetic fun expand$default (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/lang/String;Landroid/content/Context;ILjava/lang/Object;)Lcom/intuit/player/android/asset/RenderableAsset;
	public final fun expandList (Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;Landroid/content/Context;)Ljava/util/List;
	public final fun expandList (Ljava/lang/String;Landroid/content/Context;)Ljava/util/List;
	public static synthetic fun expandList$default (Lcom/intuit/player/android/asset/RenderableAsset;Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;Landroid/content/Context;ILjava/lang/Object;)Ljava/util/List;
	public static synthetic fun expandList$default (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/lang/String;Landroid/content/Context;ILjava/lang/Object;)Ljava/util/List;
	public final fun getAsset ()Lcom/intuit/player/jvm/core/asset/Asset;
	public final fun getAssetContext ()Lcom/intuit/player/android/AssetContext;
	public final fun getContext ()Landroid/content/Context;
	protected final fun getHydrationScope ()Lkotlinx/coroutines/CoroutineScope;
	public fun getNode ()Lcom/intuit/player/jvm/core/bridge/Node;
	public final fun getPlayer ()Lcom/intuit/player/android/AndroidPlayer;
	protected final fun getUnitFunction (Lcom/intuit/player/jvm/core/bridge/Node;Ljava/lang/String;)Lcom/intuit/player/jvm/core/bridge/Invokable;
	protected abstract fun hydrate (Landroid/view/View;)V
	protected abstract fun initView ()Landroid/view/View;
	public final fun invalidateView ()V
	public final fun rehydrate ()V
	public final fun render (Landroid/content/Context;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/lang/String;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/util/List;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;Ljava/util/List;Ljava/lang/String;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;[Ljava/lang/Integer;)Landroid/view/View;
	public final fun render (Lcom/intuit/player/android/asset/RenderableAsset;[Ljava/lang/Integer;Ljava/lang/String;)Landroid/view/View;
	public final fun requireContext ()Landroid/content/Context;
}

public final class com/intuit/player/android/asset/RenderableAsset$Serializer : kotlinx/serialization/KSerializer {
	public fun <init> (Lcom/intuit/player/android/AndroidPlayer;)V
	public final fun conform (Lkotlin/reflect/KClass;)Lkotlinx/serialization/KSerializer;
	public fun deserialize (Lkotlinx/serialization/encoding/Decoder;)Lcom/intuit/player/android/asset/RenderableAsset;
	public synthetic fun deserialize (Lkotlinx/serialization/encoding/Decoder;)Ljava/lang/Object;
	public fun getDescriptor ()Lkotlinx/serialization/descriptors/SerialDescriptor;
	public fun serialize (Lkotlinx/serialization/encoding/Encoder;Lcom/intuit/player/android/asset/RenderableAsset;)Ljava/lang/Void;
	public synthetic fun serialize (Lkotlinx/serialization/encoding/Encoder;Ljava/lang/Object;)V
}

public abstract interface class com/intuit/player/android/asset/RenderableAsset$ViewportAsset {
}

public final class com/intuit/player/android/databinding/DefaultFallbackBinding : androidx/viewbinding/ViewBinding {
	public final field error Landroid/widget/TextView;
	public final field reset Landroid/widget/Button;
	public final field retry Landroid/widget/Button;
	public final field title Landroid/widget/TextView;
	public static fun bind (Landroid/view/View;)Lcom/intuit/player/android/databinding/DefaultFallbackBinding;
	public synthetic fun getRoot ()Landroid/view/View;
	public fun getRoot ()Landroidx/constraintlayout/widget/ConstraintLayout;
	public static fun inflate (Landroid/view/LayoutInflater;)Lcom/intuit/player/android/databinding/DefaultFallbackBinding;
	public static fun inflate (Landroid/view/LayoutInflater;Landroid/view/ViewGroup;Z)Lcom/intuit/player/android/databinding/DefaultFallbackBinding;
}

public final class com/intuit/player/android/databinding/FragmentPlayerBinding : androidx/viewbinding/ViewBinding {
	public final field playerCanvas Landroid/widget/FrameLayout;
	public static fun bind (Landroid/view/View;)Lcom/intuit/player/android/databinding/FragmentPlayerBinding;
	public synthetic fun getRoot ()Landroid/view/View;
	public fun getRoot ()Landroid/widget/ScrollView;
	public static fun inflate (Landroid/view/LayoutInflater;)Lcom/intuit/player/android/databinding/FragmentPlayerBinding;
	public static fun inflate (Landroid/view/LayoutInflater;Landroid/view/ViewGroup;Z)Lcom/intuit/player/android/databinding/FragmentPlayerBinding;
}

public final class com/intuit/player/android/extensions/IntoKt {
	public static final fun into (Landroid/view/View;Landroid/view/ViewGroup;)V
	public static final fun into (Landroid/view/View;Landroid/widget/FrameLayout;)V
	public static final fun into (Ljava/util/List;Landroid/view/ViewGroup;)V
	public static final fun transitionInto (Landroid/view/View;Landroid/widget/FrameLayout;Landroidx/transition/Transition;)V
}

public final class com/intuit/player/android/extensions/OverlayStylesKt {
	public static final fun overlayStyles (Landroid/content/Context;Ljava/util/List;Ljava/util/List;)Landroid/content/Context;
	public static final fun overlayStyles (Landroid/content/Context;[I)Landroid/content/Context;
	public static final fun overlayStyles (Landroid/content/Context;[ILjava/lang/Integer;)Landroid/content/Context;
	public static synthetic fun overlayStyles$default (Landroid/content/Context;Ljava/util/List;Ljava/util/List;ILjava/lang/Object;)Landroid/content/Context;
}

public abstract class com/intuit/player/android/lifecycle/ManagedPlayerState {
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Done : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public fun <init> (Lcom/intuit/player/jvm/core/player/state/CompletedState;)V
	public final fun component1 ()Lcom/intuit/player/jvm/core/player/state/CompletedState;
	public final fun copy (Lcom/intuit/player/jvm/core/player/state/CompletedState;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;
	public static synthetic fun copy$default (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;Lcom/intuit/player/jvm/core/player/state/CompletedState;ILjava/lang/Object;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;
	public fun equals (Ljava/lang/Object;)Z
	public final fun getCompletedState ()Lcom/intuit/player/jvm/core/player/state/CompletedState;
	public fun hashCode ()I
	public fun toString ()Ljava/lang/String;
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Error : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public fun <init> (Ljava/lang/Exception;)V
	public final fun component1 ()Ljava/lang/Exception;
	public final fun copy (Ljava/lang/Exception;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;
	public static synthetic fun copy$default (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;Ljava/lang/Exception;ILjava/lang/Object;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;
	public fun equals (Ljava/lang/Object;)Z
	public final fun getException ()Ljava/lang/Exception;
	public fun hashCode ()I
	public fun toString ()Ljava/lang/String;
}

public abstract interface class com/intuit/player/android/lifecycle/ManagedPlayerState$Listener {
	public abstract fun onDone (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;)V
	public abstract fun onError (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;)V
	public abstract fun onNotStarted ()V
	public abstract fun onPending ()V
	public abstract fun onRunning (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;)V
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Listener$DefaultImpls {
	public static fun onDone (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;)V
	public static fun onError (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;)V
	public static fun onNotStarted (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;)V
	public static fun onPending (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;)V
	public static fun onRunning (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Listener;Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;)V
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$NotStarted : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public static final field INSTANCE Lcom/intuit/player/android/lifecycle/ManagedPlayerState$NotStarted;
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Pending : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public static final field INSTANCE Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Pending;
}

public final class com/intuit/player/android/lifecycle/ManagedPlayerState$Running : com/intuit/player/android/lifecycle/ManagedPlayerState {
	public fun <init> (Lcom/intuit/player/android/asset/RenderableAsset;Z)V
	public final fun component1 ()Lcom/intuit/player/android/asset/RenderableAsset;
	public final fun component2 ()Z
	public final fun copy (Lcom/intuit/player/android/asset/RenderableAsset;Z)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;
	public static synthetic fun copy$default (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;Lcom/intuit/player/android/asset/RenderableAsset;ZILjava/lang/Object;)Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;
	public fun equals (Ljava/lang/Object;)Z
	public final fun getAnimateViewTransition ()Z
	public final fun getAsset ()Lcom/intuit/player/android/asset/RenderableAsset;
	public fun hashCode ()I
	public fun toString ()Ljava/lang/String;
}

public class com/intuit/player/android/lifecycle/PlayerViewModel : androidx/lifecycle/ViewModel, com/intuit/player/android/AndroidPlayerPlugin, com/intuit/player/jvm/core/plugins/RuntimePlugin {
	public fun <init> (Lcom/intuit/player/jvm/core/managed/AsyncIterator;)V
	public fun apply (Lcom/intuit/player/android/AndroidPlayer;)V
	public fun apply (Lcom/intuit/player/jvm/core/bridge/runtime/Runtime;)V
	public final fun fail (Ljava/lang/Throwable;)V
	public final fun getBeacons ()Lkotlinx/coroutines/flow/SharedFlow;
	protected final fun getManager ()Lcom/intuit/player/jvm/core/managed/AsyncIterationManager;
	public final fun getPlayer ()Lcom/intuit/player/android/AndroidPlayer;
	protected fun getPlugins ()Ljava/util/List;
	public final fun getState ()Lkotlinx/coroutines/flow/StateFlow;
	public fun onCleared ()V
	public final fun recycle ()V
	public final fun release ()V
	public final fun retry ()V
	public final fun start ()V
}

public final class com/intuit/player/android/lifecycle/PlayerViewModel$Factory : androidx/lifecycle/ViewModelProvider$Factory {
	public fun <init> (Lcom/intuit/player/jvm/core/managed/AsyncIterator;Lkotlin/jvm/functions/Function1;)V
	public synthetic fun <init> (Lcom/intuit/player/jvm/core/managed/AsyncIterator;Lkotlin/jvm/functions/Function1;ILkotlin/jvm/internal/DefaultConstructorMarker;)V
	public fun create (Ljava/lang/Class;)Landroidx/lifecycle/ViewModel;
}

public final class com/intuit/player/android/lifecycle/PlayerViewModelKt {
	public static final fun fail (Lcom/intuit/player/android/lifecycle/PlayerViewModel;Ljava/lang/String;Ljava/lang/Throwable;)V
	public static synthetic fun fail$default (Lcom/intuit/player/android/lifecycle/PlayerViewModel;Ljava/lang/String;Ljava/lang/Throwable;ILjava/lang/Object;)V
}

public abstract class com/intuit/player/android/ui/PlayerFragment : androidx/fragment/app/Fragment, com/intuit/player/android/lifecycle/ManagedPlayerState$Listener {
	public fun <init> ()V
	public fun buildDoneView ()Landroid/view/View;
	public fun buildFallbackView (Ljava/lang/Exception;)Landroid/view/View;
	public fun buildLoadingView ()Landroid/view/View;
	public fun buildTransitionAnimation ()Landroidx/transition/Transition;
	protected final fun getBinding ()Lcom/intuit/player/android/databinding/FragmentPlayerBinding;
	public abstract fun getPlayerViewModel ()Lcom/intuit/player/android/lifecycle/PlayerViewModel;
	protected fun handleAssetUpdate (Lcom/intuit/player/android/asset/RenderableAsset;Z)V
	public fun onCreateView (Landroid/view/LayoutInflater;Landroid/view/ViewGroup;Landroid/os/Bundle;)Landroid/view/View;
	public fun onDestroyView ()V
	public fun onDone (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Done;)V
	public fun onError (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Error;)V
	public fun onNotStarted ()V
	public fun onPending ()V
	public fun onRunning (Lcom/intuit/player/android/lifecycle/ManagedPlayerState$Running;)V
	public final fun reset ()V
}

