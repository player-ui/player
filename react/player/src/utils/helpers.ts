/**
 * Trim leading and trailing slashes from string
 */
export function trimSlashes(str: string) {
  return str.replace(/^\/+|\/+$/g, '');
}

/**
 * Removes any key: value pairs from an object when the value is null or undefined
 */
export function removeEmptyValuesFromObject(
  obj: Record<string, any>
): Record<string, NonNullable<any>> {
  return Object.keys(obj).reduce((acc, key) => {
    const value = obj[key];

    if (value !== null && value !== undefined) {
      acc[key] = value;
    }

    return acc;
  }, {} as Record<string, any>);
}

/** Check if the object has no keys */
export function isEmptyObject(obj: Record<string, unknown>) {
  return Object.keys(obj).length === 0 && obj.constructor === Object;
}

/** Check if the argument is a function */
export function isFunction<ReturnType>(
  maybeFn: ReturnType | ((...args: unknown[]) => ReturnType)
): maybeFn is (...args: unknown[]) => ReturnType {
  return Boolean(maybeFn instanceof Function || typeof maybeFn === 'function');
}

/**
 * Calls function with provided data or returns original value
 */
export function callOrReturn<
  ReturnType,
  FnArgs extends Array<unknown> = unknown[],
  FnType = (...args: FnArgs) => ReturnType
>(maybeFn: FnType | ReturnType, fnArgs: FnArgs): ReturnType {
  if (isFunction(maybeFn)) {
    return maybeFn(fnArgs) as ReturnType;
  }

  return maybeFn as ReturnType;
}
