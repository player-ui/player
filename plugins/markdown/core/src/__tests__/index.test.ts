import type { InProgressState } from '@player-ui/player';
import { Player } from '@player-ui/player';
import { Registry } from '@player-ui/partial-match-registry';
import { PartialMatchFingerprintPlugin } from '@player-ui/partial-match-fingerprint-plugin';
import type { Flow } from '@player-ui/types';
import { mockMappers } from './helpers';
import { MarkdownPlugin } from '..';

const unparsedFlow: Flow = {
  id: 'markdown-flow',
  data: {
    internal: {
      locale: {
        linkMarkdown:
          'Learn more at [TurboTax Canada](https://turbotax.intuit.ca)',
      },
    },
  },
  views: [
    {
      id: 'markdown-view',
      type: 'questionAnswer',
      title: {
        asset: {
          id: 'markdown-view-title',
          type: 'markdown',
          value: '{{internal.locale.linkMarkdown}}',
        },
      },
      primaryInfo: {
        asset: {
          id: 'markdown-primaryInfo-collection',
          type: 'collection',
          values: [
            {
              asset: {
                id: 'markdown-primaryInfo-collection-bold',
                type: 'markdown',
                value: 'some **bold text**',
              },
            },
            {
              asset: {
                id: 'markdown-primaryInfo-collection-italic',
                type: 'markdown',
                value: '*italicized text*',
              },
            },
            {
              asset: {
                id: 'markdown-primaryInfo-collection-orderd-list',
                type: 'markdown',
                value: '1. First\n2. Second\n3. Third',
              },
            },
            {
              asset: {
                id: 'markdown-primaryInfo-collection-unorderd-list',
                type: 'markdown',
                value:
                  '- [First](https://turbotax.intuit.ca)\n- Second\n- Third',
              },
            },
            {
              asset: {
                id: 'markdown-primaryInfo-collection-image',
                type: 'markdown',
                value: '![alt text](image.png)',
              },
            },
            {
              asset: {
                id: 'markdown-primaryInfo-collection-unsupported',
                type: 'markdown',
                value: 'Highlights are ==not supported==',
              },
            },
          ],
        },
      },
    },
  ],
  navigation: {
    BEGIN: 'FLOW_1',
    FLOW_1: {
      startState: 'VIEW_1',
      VIEW_1: {
        state_type: 'VIEW',
        ref: 'markdown-view',
        transitions: {
          '*': 'END_Done',
        },
      },
      END_Done: {
        state_type: 'END',
        outcome: 'done',
      },
    },
  },
};

describe('MarkdownPlugin', () => {
  it('parses the flow containing markdown into valid FRF, based on the given mappers', () => {
    const player = new Player({
      plugins: [new MarkdownPlugin(mockMappers)],
    });
    player.start(unparsedFlow);

    const view = (player.getState() as InProgressState).controllers.view
      .currentView?.lastUpdate;

    expect(view).toMatchSnapshot();
  });

  it('parses the flow, with only the required mappers', () => {
    const player = new Player({
      plugins: [
        new MarkdownPlugin({
          text: mockMappers.text,
          paragraph: mockMappers.paragraph,
        }),
      ],
    });
    player.start(unparsedFlow);

    const view = (player.getState() as InProgressState).controllers.view
      .currentView?.lastUpdate;

    expect(view).toMatchSnapshot();
  });

  it('parses regular flow and maps assets', () => {
    const fingerprint = new PartialMatchFingerprintPlugin(new Registry());

    fingerprint.register({ type: 'action' }, 0);
    fingerprint.register({ type: 'text' }, 1);
    fingerprint.register({ type: 'composite' }, 2);

    const player = new Player({
      plugins: [fingerprint, new MarkdownPlugin(mockMappers)],
    });

    player.start({
      id: 'action-with-expression',
      views: [
        {
          id: 'action',
          type: 'action',
          exp: '{{count}} = {{count}} + 1',
          label: {
            asset: {
              id: 'action-label',
              type: 'markdown',
              value: 'Clicked {{count}} *times*',
            },
          },
        },
      ],
      data: {
        count: 0,
      },
      navigation: {
        BEGIN: 'FLOW_1',
        FLOW_1: {
          startState: 'VIEW_1',
          VIEW_1: {
            state_type: 'VIEW',
            ref: 'action',
            transitions: {
              '*': 'END_Done',
            },
          },
          END_Done: {
            state_type: 'END',
            outcome: 'done',
          },
        },
      },
    });

    // the parser should create 2 text assets: `Clicked {{count}}` and a italicized `times`:
    expect(fingerprint.get('action-label-text-38')).toBe(1);
    expect(fingerprint.get('action-label-text-39')).toBe(1);
  });
});
