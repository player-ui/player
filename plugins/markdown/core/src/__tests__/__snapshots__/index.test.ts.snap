// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownPlugin parses the flow containing markdown into valid FRF, based on the given mappers 1`] = `
Object {
  "id": "markdown-view",
  "primaryInfo": Object {
    "asset": Object {
      "id": "markdown-primaryInfo-collection",
      "type": "collection",
      "values": Array [
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-bold-composite-7",
            "type": "composite",
            "values": Array [
              Object {
                "asset": Object {
                  "id": "markdown-primaryInfo-collection-bold-text-4",
                  "type": "text",
                  "value": "some ",
                },
              },
              Object {
                "asset": Object {
                  "id": "markdown-primaryInfo-collection-bold-text-5",
                  "modifiers": Array [
                    Object {
                      "type": "tag",
                      "value": "important",
                    },
                  ],
                  "type": "text",
                  "value": "bold text",
                },
              },
            ],
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-italic-text-8",
            "modifiers": Array [
              Object {
                "type": "tag",
                "value": "emphasis",
              },
            ],
            "type": "text",
            "value": "italicized text",
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-orderd-list-list-20",
            "metaData": Object {
              "listType": "ordered",
            },
            "type": "list",
            "values": Array [
              Object {
                "asset": Object {
                  "id": "markdown-primaryInfo-collection-orderd-list-text-11",
                  "type": "text",
                  "value": "First",
                },
              },
              Object {
                "asset": Object {
                  "id": "markdown-primaryInfo-collection-orderd-list-text-14",
                  "type": "text",
                  "value": "Second",
                },
              },
              Object {
                "asset": Object {
                  "id": "markdown-primaryInfo-collection-orderd-list-text-17",
                  "type": "text",
                  "value": "Third",
                },
              },
            ],
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-unorderd-list-list-31",
            "type": "list",
            "values": Array [
              Object {
                "asset": Object {
                  "id": "markdown-primaryInfo-collection-unorderd-list-text-21",
                  "modifiers": Array [
                    Object {
                      "metaData": Object {
                        "ref": "https://turbotax.intuit.ca",
                      },
                      "type": "link",
                    },
                  ],
                  "type": "text",
                  "value": "First",
                },
              },
              Object {
                "asset": Object {
                  "id": "markdown-primaryInfo-collection-unorderd-list-text-25",
                  "type": "text",
                  "value": "Second",
                },
              },
              Object {
                "asset": Object {
                  "id": "markdown-primaryInfo-collection-unorderd-list-text-28",
                  "type": "text",
                  "value": "Third",
                },
              },
            ],
          },
        },
        Object {
          "asset": Object {
            "accessibility": "alt text",
            "id": "markdown-primaryInfo-collection-image-image-32",
            "metaData": Object {
              "ref": "image.png",
            },
            "type": "image",
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-unsupported-text-34",
            "type": "text",
            "value": "Highlights are ==not supported==",
          },
        },
      ],
    },
  },
  "title": Object {
    "asset": Object {
      "id": "markdown-view-title-composite-3",
      "type": "composite",
      "values": Array [
        Object {
          "asset": Object {
            "id": "markdown-view-title-text-0",
            "type": "text",
            "value": "Learn more at ",
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-view-title-text-1",
            "modifiers": Array [
              Object {
                "metaData": Object {
                  "ref": "https://turbotax.intuit.ca",
                },
                "type": "link",
              },
            ],
            "type": "text",
            "value": "TurboTax Canada",
          },
        },
      ],
    },
  },
  "type": "questionAnswer",
}
`;

exports[`MarkdownPlugin parses the flow, with only the required mappers 1`] = `
Object {
  "id": "markdown-view",
  "primaryInfo": Object {
    "asset": Object {
      "id": "markdown-primaryInfo-collection",
      "type": "collection",
      "values": Array [
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-bold",
            "type": "text",
            "value": "some **bold text**",
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-italic",
            "type": "text",
            "value": "*italicized text*",
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-orderd-list",
            "type": "text",
            "value": "1. First
2. Second
3. Third",
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-unorderd-list",
            "type": "text",
            "value": "- [First](https://turbotax.intuit.ca)
- Second
- Third",
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-image",
            "type": "text",
            "value": "![alt text](image.png)",
          },
        },
        Object {
          "asset": Object {
            "id": "markdown-primaryInfo-collection-unsupported-text-36",
            "type": "text",
            "value": "Highlights are ==not supported==",
          },
        },
      ],
    },
  },
  "title": Object {
    "asset": Object {
      "id": "markdown-view-title",
      "type": "text",
      "value": "Learn more at [TurboTax Canada](https://turbotax.intuit.ca)",
    },
  },
  "type": "questionAnswer",
}
`;
